### PR TITLE
New consensus event type for currency messages

### PR DESCRIPTION
**Pull request by marcinwadon**

### CODE REVIEW

File: modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/CurrencySnapshotAcceptanceManager.scala [hunk 1]
```diff
@@ -74,10 +91,48 @@ object CurrencySnapshotAcceptanceManager {
...
+
+      messagesToReject = messagesForAcceptance.filterNot(messagesToAccept.values.toList.flatten.contains)
+
+      messagesForContextUpdate = (lastMessages.toList ++ messagesToAccept.toList.flatMap {
+        case (t, xs) => xs.map(x => (t, x))
+      }).groupBy { case ((messageType, _)) => messageType }.view.mapValues {
+        _.map { case ((_, messages)) => messages }.maxByOption(_.ordinal)
+      }.toMap.collect {
+        case (k, Some(v)) => k -> v
+      }.toSortedMap
```
Comment (sdavidp) [inline modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/CurrencySnapshotAcceptanceManager.scala:124]: The formatting makes this a bit hard to follow but is this refactor equivalent to what we're trying to do here?
```
messagesForContextUpdate =
  messagesToAccept
    .view
    .mapValues(_.maxByOption(_.ordinal))
    .foldLeft(lastMessages) {
      case (acc, (k, v)) =>
        acc.updatedWith(k) {
          case None => v
          case x@Some(v0) => v.filter(_.ordinal > v0.ordinal).orElse(x)
        }
    }
```

File: modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/CurrencySnapshotAcceptanceManager.scala [hunk 2]
```diff
@@ -74,10 +91,48 @@ object CurrencySnapshotAcceptanceManager {
...
+        .groupBy(_.messageType)
+        .view
+        .mapValues { messages =>
+          val sortedMessages = messages.sortBy(_.parentOrdinal)
+
+          sortedMessages
+            .foldLeft(List.empty[Signed[CurrencyMessage]]) { (acc, curr) =>
+              lastMessages.get(curr.messageType) match {
+                case Some(lastMsg) if curr.parentOrdinal <= lastMsg.parentOrdinal => acc
+                case _ if acc.isEmpty || acc.last.ordinal == curr.parentOrdinal   => curr :: acc
```
[the remark this answers is not in the document]
Comment (marcinwadon) [reply modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/CurrencySnapshotAcceptanceManager.scala:106]: what's the preferred way then? wait. actually we're prepending, so we should take acc.head, shouldn't we?

[the remark this answers is not in the document]
Comment (marcinwadon) [reply modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/CurrencySnapshotAcceptanceManager.scala:106]: yep, we should take the `acc.head`

File: modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/CurrencySnapshotAcceptanceManager.scala (hunk 1 above)
Comment (marcinwadon) [reply modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/CurrencySnapshotAcceptanceManager.scala:124]: well, I do the following steps:
1. group by MessageType
2. take a single message from the list by the max ordinal
3. filter out None (if there is no message for the message type

File: modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/CurrencySnapshotAcceptanceManager.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (marcinwadon) [reply modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/CurrencySnapshotAcceptanceManager.scala]: fixed

[the remark this answers is not in the document]
Comment (sdavidp) [reply modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/CurrencySnapshotAcceptanceManager.scala]: Cool, I wondered about that.

File: modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/CurrencySnapshotAcceptanceManager.scala (hunk 1 above)
Comment (sdavidp) [reply modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/CurrencySnapshotAcceptanceManager.scala:124]: Ok, so I think I understood the intent. In that case, isn't the above refactor a simplification? In effect,

1. For all messages to accept, find the max message for each message type.
2. Merge the result with `lastMessages`

The result from step 1 has the same type as `lastMessages` so I think we can avoid most of the intermediary data structures.

Comment (marcinwadon) [reply modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/CurrencySnapshotAcceptanceManager.scala:124]: Working on immutable data structures, we create new instances anyway. Does it really matter it's this or that type as an intermediate type? I think we need to select between readability and efficiency and in this case we simply follow the easy algorithm and the second case requires an additional logic for merging both structures (instead of merging first and then operating on all messages) 🤔

File: modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/CurrencySnapshotAcceptanceManager.scala [hunk 3]
```diff
@@ -74,10 +91,48 @@ object CurrencySnapshotAcceptanceManager {
...
+                case _                                                            => acc
+              }
+            }
+        }
+        .toMap
+        .collect {
+          case (k, x :: xs) => (k, x :: xs)
+        }
```
[the remark this answers is not in the document]
Comment (marcinwadon) [reply modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/CurrencySnapshotAcceptanceManager.scala:114]: :+1:

File: modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/CurrencySnapshotAcceptanceManager.scala (hunk 1 above)
Comment (marcinwadon) [reply modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/CurrencySnapshotAcceptanceManager.scala]: Ok. I see I can use a semigroup to make it easier and more readable:
1. merge 
2. take highest by ordinal
3. filter-out None

image

**Review by sdavidp — APPROVED** [review]
